tests/std/tests/LWG2510_tag_classes/test.compile.pass.cpp

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstddef>
+#include <functional>
+#include <list>
+#include <memory>
+#include <new>
+#include <scoped_allocator>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#ifndef _M_CEE_PURE
+#include <future>
+#include <mutex>
+#endif // _M_CEE_PURE
+
+#if _HAS_CXX17
+#include <execution>
+#endif // _HAS_CXX17
+
+#if _HAS_CXX20
+#include <chrono>
+#include <format>
+#include <iterator>
+#include <ranges>
+#include <stop_token>
+#endif // _HAS_CXX20
+
+#if _HAS_CXX23
+#include <expected>
+#include <generator>
+#include <mdspan>
+#include <optional>
+#endif // _HAS_CXX23
+
+#define STATIC_ASSERT(...) static_assert(__VA_ARGS__, #__VA_ARGS__)
+
+using namespace std;
+
+// LWG-2510 "Tag types should not be DefaultConstructible"
+
+template <class T>
+void check_implicit_default_construction(const T&); // not defined
+
+template <class T, class = void>
+constexpr bool implicitly_default_constructible = false;
+template <class T>
+constexpr bool implicitly_default_constructible<T, void_t<decltype(check_implicit_default_construction<T>({}))>> = true;
+
+STATIC_ASSERT(implicitly_default_constructible<int>);
+STATIC_ASSERT(implicitly_default_constructible<void*>);
+
+STATIC_ASSERT(!implicitly_default_constructible<void>);
+STATIC_ASSERT(!implicitly_default_constructible<int&>);
+STATIC_ASSERT(!implicitly_default_constructible<void()>);
+
+template <class T>
+void check_tag_class() { // COMPILE-ONLY
+    STATIC_ASSERT(is_default_constructible_v<T>);
+    STATIC_ASSERT(!implicitly_default_constructible<T>);
+
+    STATIC_ASSERT(is_copy_constructible_v<T>);
+    STATIC_ASSERT(is_move_constructible_v<T>);
+    STATIC_ASSERT(is_constructible_v<T, T&>);
+    STATIC_ASSERT(is_constructible_v<T, const T>);

```

Comment:
No change requested: It took me a while to realize that `is_copy_constructible_v` and `is_move_constructible_v` were covering 2 cases and `is_constructible_v` were covering the other 2 cases.